### PR TITLE
Update/gds performance imrovement

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Originating Project/Creator_**| |
|**_Affected Component_**|  |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Builds Without Errors (y/n)_**|  |
|**_Unit Tests Pass (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

Parallel changes to the fprime-tools rework.

### CODE REVIEW

File: src/fprime_gds/common/data_types/cmd_data.py [hunk 1]
```diff
@@ -157,6 +147,28 @@ def get_str(self, time_zone=None, verbose=False, csv=False):
...
+        for val, arg_tuple in zip(input_values, self.template.arguments):
+            try:
+                _, _, arg_type = arg_tuple
+                arg_value = arg_type()
+                self.convert_arg_value(val, arg_value)
+                args.append(arg_value)
+                errors.append("")
+            except Exception as exc:
+                error_message = str(exc)
+                # Patch old versions of fprime-tools to replace a bad error message with the correct one
```
Comment (kevin-f-ortega) [inline src/fprime_gds/common/data_types/cmd_data.py]: This is dead code. Please remove it.

File: src/fprime_gds/common/decoders/decoder.py [hunk 2]
```diff
@@ -46,7 +50,13 @@ def data_callback(self, data, sender=None):
         :param data: data bytes to be decoded
         :param sender: (optional) sender id, otherwise None
         """
-        decoded = self.decode_api(data)
+        try:
+            decoded = self.decode_api(data)
+        except DecodingException:
+            raise
```
Comment (kevin-f-ortega) [inline src/fprime_gds/common/decoders/decoder.py:57]: Add a comment about why you are just raising exception here.

File: src/fprime_gds/common/distributor/distributor.py [hunk 3]
```diff
@@ -97,27 +103,24 @@ def parse_into_raw_msgs_api(self, data):
                     # Check leading key size bytes to see if it is the key
                     self.key_obj.deserialize(data_left, 0)
                     if self.key_obj.val != self.key_frame:
-                        data_left = data_left[1:]
+                        del data_left[1]
```
Comment (kevin-f-ortega) [inline src/fprime_gds/common/distributor/distributor.py]: Should be `del data_left[0]`

File: src/fprime_gds/common/distributor/distributor.py [hunk 4]
```diff
@@ -180,15 +190,18 @@ def on_recv(self, data):
 
         # Add new data to end of buffer
 
-        self.__buf = self.__buf + data
+        self.__buf.extend(data)
 
         (leftover_data, raw_msgs) = self.parse_into_raw_msgs_api(self.__buf)
-        self.__buf = leftover_data
+        assert leftover_data == self.__buf, "Weirdness happened"
```
Comment (kevin-f-ortega) [inline src/fprime_gds/common/distributor/distributor.py]: Assert message should be more descriptive.

File: src/fprime_gds/flask/json.py [hunk 5]
```diff
@@ -3,16 +3,80 @@
...
+        class properties
+    }
+
+    Args:
+        input_type: input to convert to dictionary
+    Return:
+        json-able dictionary representing the type
+    """
+    assert issubclass(input_type, BaseType), "Failure to properly encode data"
+    members = getmembers(input_type, lambda value: not isroutine(value) and not isinstance(value, property))
```
Comment (kevin-f-ortega) [inline src/fprime_gds/flask/json.py:39]: Is this correct?

File: src/fprime_gds/flask/json.py [hunk 6]
```diff
@@ -3,16 +3,80 @@
...
+
 class GDSJsonEncoder(flask.json.JSONEncoder):
     """
     Custom class used to handle GDS object to JSON
     """
+    JSON_ENCODERS = {
+        ABCMeta: jsonify_base_type,
+        UUID: str,
+        ChData: minimal_channel,
+        EventData: minimal_event
```
Comment (kevin-f-ortega) [inline src/fprime_gds/flask/json.py]: Add a command encoder here.

File: src/fprime_gds/flask/json.py [hunk 7]
```diff
@@ -3,16 +3,80 @@
...
+from typing import Type
 
+from fprime.common.models.serialize.type_base import BaseType
+from fprime_gds.common.data_types.ch_data import ChData
+from fprime_gds.common.data_types.event_data import EventData
+from uuid import UUID
 import flask.json
 
 
+def jsonify_base_type(input_type: Type[BaseType]) -> dict:
```
Comment (kevin-f-ortega) [inline src/fprime_gds/flask/json.py:22]: Is .val appropriate here?

File: src/fprime_gds/flask/static/addons/chart-display/addon.js [hunk 8]
```diff
@@ -190,25 +227,44 @@ Vue.component("chart-display", {
...
+                this.oldSelected = new_selected;
+                this.$nextTick().then(() => {this.registerChart()});
+            }
+        }
     },
     /**
      * Watch for new selection of channel and re-register the chart
      */
-    watch: {
+    /**watch: {
```
Comment (kevin-f-ortega) [inline src/fprime_gds/flask/static/addons/chart-display/addon.js]: Deadcode

File: src/fprime_gds/flask/static/addons/chart-display/modified-vendor/flat.js [hunk 9]
```diff
@@ -29,7 +29,7 @@ function flatten(target, opts) {
     opts = opts || {};
     // Default supported F Prime types
     const supportedTypes =
-        ["U8", "U16", "U32", "U64", "I8", "I16", "I32", "I64", "F32", "F64"] ||
+        ["U8Type", "U16Type", "U32Type", "U64Type", "I8Type", "I16Type", "I32Type", "I64Type", "F32Type", "F64Type"] ||
```
Comment (kevin-f-ortega) [inline src/fprime_gds/flask/static/addons/chart-display/modified-vendor/flat.js]: Include serializable

File: src/fprime_gds/flask/static/js/config.js [hunk 10]
```diff
@@ -20,5 +20,10 @@ export let config = {
     dataPollIntervalsMs: {
         channels: 500,
         default: 1000
-    }
+    },
+    // Summary counter fields containing object of field: bootstrap class
+    summaryFields: {"WARNING_HI": "warning", "FATAL": "danger", "GDS_Errors": "danger"},
+    // Allow projects to configure what tabs will be used. All tabs default to enabled.
+    // Disable tabs using: <tab lowercase fullname: false> e.g. logs: false" in the mapping below
+    enabledTabs: {}
```
Comment (kevin-f-ortega) [inline src/fprime_gds/flask/static/js/config.js]: Does this work?

File: src/fprime_gds/flask/static/js/vue-support/tabetc.js [hunk 11]
```diff
@@ -79,6 +82,15 @@ Vue.component("tabbed-etc", {
...
                 } 
             }
+        },
+        /**
+         * Check if the supplied tab is enabled w.r.t the configuration of the GUI.
+         * @param tab: tab fullname to check against
+         */
+        tabEnabled(tab) {
+            return true;
+            /*tab = tab.toLowerCase();
```
Comment (kevin-f-ortega) [inline src/fprime_gds/flask/static/js/vue-support/tabetc.js]: deadcode?

**Review by kevin-f-ortega — COMMENTED** [review]
Good job! Looks good. Some comments.

File: src/fprime_gds/common/distributor/distributor.py (hunk 3 above)
Comment (LeStarch) [reply src/fprime_gds/common/distributor/distributor.py]: Great catch!

File: src/fprime_gds/flask/static/addons/chart-display/modified-vendor/flat.js (hunk 9 above)
Comment (LeStarch) [reply src/fprime_gds/flask/static/addons/chart-display/modified-vendor/flat.js]: It is arrays that are failing here.

Comment (LeStarch) [reply src/fprime_gds/flask/static/addons/chart-display/modified-vendor/flat.js]: Turns out it was arrays failing.  This is now fixed.

**Comment by LeStarch** [comment]
@kevin-f-ortega this is ready once CI passes

**Review by kevin-f-ortega — APPROVED** [review]
This looks good to me. Great work!

**Pull request merged**